Foundation/NSKeyedUnarchiver.swift

```diff
@@ -0,0 +1,955 @@
+        }
+        
+        let version = unwrappedPlist["$version"] as? NSNumber
+        if version?.intValue != Int32(NSKeyedArchivePlistVersion) {
+            throw _decodingError(NSCocoaError.PropertyListReadCorruptError,
+                                 withDescription: "Unknown archive version. The data may be corrupt.")
+        }
+        
+        let top = unwrappedPlist["$top"] as? Dictionary<String, Any>
+        let objects = unwrappedPlist["$objects"] as? Array<Any>
+        
+        if top == nil || objects == nil {
+            throw _decodingError(NSCocoaError.PropertyListReadCorruptError,
+                                 withDescription: "Unable to read archive contents. The data may be corrupt.")
+        }
+        
+        self._objects = objects!
+        self._containers = [DecodingContext(top!)]
+    }
+    
+    private func _pushDecodingContext(decodingContext: DecodingContext) {
+        self._containers!.append(decodingContext)
+    }
+    
+    private func _popDecodingContext() {
+        self._containers!.removeLast()
+    }
+    
+    private var _currentDecodingContext : DecodingContext {
+        return self._containers!.last!
+    }
+    
+    private func _nextGenericKey() -> String {
+        let key = "$" + String(_currentDecodingContext.genericKey)
+        _currentDecodingContext.genericKey += 1
+        return key
+    }
+    
+    private func _objectInCurrentDecodingContext<T>(forKey key: String?) -> T? {
+        var unwrappedKey = key
+        
+        if key != nil {
+            unwrappedKey = escapeArchiverKey(key!)
+        } else {
+            unwrappedKey = _nextGenericKey()
+        }
+
+        return _currentDecodingContext.dict[unwrappedKey!] as? T
+    }
+    
+    /**
+        Dereferences, but does not decode, an object reference
+     */
+    private func _dereferenceObjectReference(unwrappedObjectRef: CFKeyedArchiverUID) -> Any? {
+        let uid = Int(objectRefGetValue(unwrappedObjectRef))
+        
+        guard uid < self._objects.count else {
+            return nil
+        }
+
+        return self._objects[uid]
+    }
+    
+    public override var systemVersion: UInt32 {
+        return NSKeyedArchiverSystemVersion
+    }
+    
+    public override var allowsKeyedCoding: Bool {
+        get {
+            return true
+        }
+    }
+    
+    private func _validateStillDecoding() -> Bool {
+        if self._flags.contains(UnarchiverFlags.FinishedDecoding) {
+            fatalError("Decoder already finished")
+        }
+        
+        return true
+    }
+    
+    private static func _supportsSecureCoding(clsv : AnyClass) -> Bool {
+        if let secureCodable = clsv as? NSSecureCoding.Type {
+            return secureCodable.supportsSecureCoding()
+        }
+        
+        return false
+    }
+    
+    // FIXME is there a better way to do this with Swift stdlib?
```

Comment:
Is there a way to do a dynamic cast given types (rather than instances) in the language itself? We could call swift_dynamicCastMetatype() but that's not public.